KAFKA-7514: Add threads to ConsumeBenchWorker

stanislavkozlovski opened the pull request:
This PR adds a new ConsumeBenchSpec field - "consumerCount". "consumerCount" will be spawned over "consumerCount" threads in the ConsumeBenchWorker.
Since "consumerCount" will be 1 by default, these changes are backwards compatible

It's now questionable how existing fields such as "targetMessagesPerSec", "maxMessages", "consumerGroup" and "activeTopics" should work.

With "activeTopics", we need to decide whether they should be split over the consumers or not.
I see 4 cases which I believe we should address like this:

1. Random group, subscribe to topics - N unique groups all subscribed to all topics
2. Specifed group, subscribe to topics - 1 group subscribed to all topics. Consumers share workload.
3. Random groups, assign partitions - X groups all subscribed to all partitions
4. Specified group, assign partitions - Throw an exception. Only one consumer can read from a specific partition within the context of a consumer group. It is then unclear how and whether at all we should split the partitions across the consumers. At this phase, I believe it's best to not support this.


I believe "targetMessagesPerSec", "maxMessages" should account for each consumer individually. This would ease implementation by a ton, too.


I haven't written tests yet since I want to flesh out the design first. Any feedback is appreciated

stanislavkozlovski committed:
KAFKA-7514: Add threads to ConsumeBenchWorker

stanislavkozlovski committed:
Add status updaters for the separate consumers

stanislavkozlovski committed:
Ensure ConsumeBench does not consume more than max messages

Also call consumer#unsubscribe() and consumer#close() on finishing the task

stanislavkozlovski committed:
Do not close ConsumeBenchWorker on a single consumer status update failure

Abort the worker if every status updater fails

stanislavkozlovski committed:
Add assignedPartitions field to consumer status

Also bundle up clientId and KafkaConsumer into a single class

stanislavkozlovski committed:
Ensure thread-safe access to KafkaConsumer by wrapping operations in private Consumer class

stanislavkozlovski committed:
Make Consumer#close() idempotent

stanislavkozlovski committed:
Update tests

stanislavkozlovski committed:
Update Trogdor test scripts to use new functionality

stanislavkozlovski committed:
Update documentation

stanislavkozlovski committed:
Handle undefined `targetMessagesPerSec` in ConsumeBenchSpec

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -182,9 +237,11 @@ public Void call() throws Exception {
-                    ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(50));
+                while (messagesConsumed < maxMessages) {
+
+                    ConsumerRecords<byte[], byte[]> records = consumer.poll();
```
This raises an `InterruptException` and I can't figure out why. Is there something obvious that I'm missing?
```
[2018-11-05 17:47:05,859] WARN ConsumeRecords caught an exception:  (org.apache.kafka.trogdor.workload.ConsumeBenchWorker)
org.apache.kafka.common.errors.InterruptException: java.lang.InterruptedException
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.maybeThrowInterruptException(ConsumerNetworkClient.java:493)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:281)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:236)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1243)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1188)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1164)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$Consumer.poll(ConsumeBenchWorker.java:503)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$ConsumeMessages.call(ConsumeBenchWorker.java:243)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$ConsumeMessages.call(ConsumeBenchWorker.java:198)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.InterruptedException
	... 14 more
```

cmccabe replied:
Most likely something somewhere else failed, and your thread was sent the interrupt exception as part of tearing down the worker.

stanislavkozlovski replied:
I am not sure about the cause of this since I haven't been able to reproduce it since. I had changed the parameters in `trogdor-run-consume-bench.sh`. Might have been something to do with old consumer group state after having re-ran it a couple of times

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -202,62 +259,140 @@ public Void call() throws Exception {
                     startBatchMs = Time.SYSTEM.milliseconds();
                 }
             } catch (Exception e) {
+                // TODO: Should we close task on consumer failure?
```
This is an open question - I'm not sure what's better - to abort the whole task if one consumer fails or continue running until the last one fails

cmccabe replied:
Please abort the whole task if the consumer fails.  That's how the rest of the tests work.

In general, the consumer should never fail-- if it does, there is a serious problem.

stanislavkozlovski replied:
Makes sense. I have logic implemented that does not fail the whole task is a `StatusUpdater` of one consumer fails - only when all `StatusUpdaters` fail does the task get aborted. Thoughts on that?

cmccabe replied:
The status updaters should not fail, right?  Any failure should abort the worker.

stanislavkozlovski replied:
Seems reasonable. I reverted my commit

cmccabe commented:
Please let's use "thread count" instead of "consumer count."  We may eventually want to extend this job to be able to spawn multiple workers (if it isn't already).

I think the questions above about how max messages per sec, etc. work are resolved, right?  They apply on a per-thread basis.  We should add some JavaDoc specifying this.

stanislavkozlovski commented:
Thanks for the review @cmccabe. The questions on maxMessages and etc are resolves, yes. I have already added a JavaDoc explaining this in `ConsumeBenchSpec.java` - please take a look to see if they are adequate

stanislavkozlovski committed:
Rename consumeCount->threadCount

stanislavkozlovski committed:
Stylistic changes

stanislavkozlovski committed:
Revert "Do not close ConsumeBenchWorker on a single consumer status update failure"

This reverts commit 63ebaa9603258e37f1153398415899819aa1ad58.

stanislavkozlovski committed:
Rename inner Consumer class and change to static inner class (findbugs warning)

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java:
```diff
@@ -100,6 +108,7 @@ public ConsumeBenchSpec(@JsonProperty("startMs") long startMs,
                             @JsonProperty("consumerConf") Map<String, String> consumerConf,
                             @JsonProperty("commonClientConf") Map<String, String> commonClientConf,
                             @JsonProperty("adminClientConf") Map<String, String> adminClientConf,
+                            @JsonProperty("threadCount") Integer threadCount,
```
Can we call this "threadsPerWorker" instead?  We will probably want to support multiple workers in the future.

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java:
```diff
@@ -138,6 +148,11 @@ public int maxMessages() {
 
+    @JsonProperty
+    public int consumerCount() {
+        return consumerCount;
```
We should make this consistent with the name of the varialbe in the spec

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java:
```diff
@@ -61,6 +62,12 @@
+ * It is worth noting that the "targetMessagesPerSec", "maxMessages" and "activeTopics" fields apply for every consumer individually.
+ * If a consumer group is not specified, every consumer is assigned a different, random group. When specified, all consumers use the same group.
+ * Specifying partitions, a consumer group and multiple consumers will result in an #{@link ConfigException} and the task will abort.
+ *
```
It would probably be good to have a separate paragraph to describe the group configuration.

We probably want to mention that if individual partitions are specified, only a single consumer is supported, and subscribe is not used.

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -77,9 +85,12 @@ public void start(Platform platform, WorkerStatusTracker status,
-        this.status = status;
+            spec.consumerCount() + 2, // 1 thread for all the ConsumeStatusUpdater and 1 for the StatusUpdater
+            ThreadUtils.createThreadFactory("ConsumeBenchWorkerThread%d", false));
+        this.statusUpdaterFuture = executor.scheduleAtFixedRate(this.statusUpdater, 2, 1, TimeUnit.MINUTES);
```
Hmm.  Why change the initial delay from 1 minute to 2?

stanislavkozlovski replied:
Whoops, I initially thought this was seconds. My reasoning was that since there are more threads spawned and a rebalance happening, it would make sense to wait a bit more before exposing the status.
Anyway, reverting this back to 1 minute

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
+            boolean toUseGroupPartitionAssignment = partitionsByTopic.values().stream().allMatch(List::isEmpty);
+
+            if (!toUseGroupPartitionAssignment && !toUseRandomConsumeGroup() && consumerCount > 1)
+                throw new ConfigException(String.format("Will not split partitions across consumers from the %s group", consumerGroup));
```
This exception seems a bit confusing.

Maybe something like "You may not specify an explicit partition assignment when using multiple consumers in the same group.  Please either leave the consumer group unset, or specify topics instead of partitions."

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
+
+            String clientId = clientId(0);
+            consumer = consumer(consumerGroup, clientId);
+            if (!toUseGroupPartitionAssignment) {
```
Since we're handling both the case where this is true, and where it is false, we might as well test for (toUseGroupPartitionAssignment) rather than (!toUseGroupPartitionAssignment)

stanislavkozlovski replied:
True, that's more readable

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
+         */
+        private ThreadSafeConsumer consumer(String consumerGroup, String idx) {
             Properties props = new Properties();
+            String clientId = String.format("consumer.%s-%s", id, idx);
```
This seems to duplicate the code in the `clientId(int idx)` function?

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
-        private String generateConsumerGroup() {
-            return "consume-bench-" + UUID.randomUUID().toString();
+        private boolean toUseRandomConsumeGroup() {
+            return spec.consumerGroup().equals(ConsumeBenchSpec.EMPTY_CONSUMER_GROUP);
```
You should be able to just say `spec.consumerGroup().isEmpty()`

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -151,29 +199,35 @@ private String generateConsumerGroup() {
-            }
+                new ConsumeStatusUpdater(latencyHistogram, messageSizeHistogram, consumer), 1, 1, TimeUnit.MINUTES);
+            int perPeriod;
+            if (spec.targetMessagesPerSec() == 0)
```
Suggest using <= 0

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -202,6 +257,9 @@ public Void call() throws Exception {
                         latencyHistogram.add(elapsedBatchMs);
                         messageSizeHistogram.add(messageBytes);
                         bytesConsumed += messageBytes;
+                        if (messagesConsumed >= maxMessages)
```
Why is this needed?  The loop condition is the same.

stanislavkozlovski replied:
The default `max.poll.records` is 500. This would result in the consumer always consuming 500 at a minimum (provided the topic had the records) and not respect the cases where `maxMessages` was less than that

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -333,11 +452,86 @@ public void stop(Platform platform) throws Exception {
+    /**
+     * A thread-safe KafkaConsumer wrapper
+     */
+    private static class ThreadSafeConsumer {
```
This is not needed.  Only one thread at a time should be accessing the consumer.

stanislavkozlovski replied:
The `ConsumeStatusUpdater`  calls `KafkaConsumer#assignment()` dynamically to get the latest assignments for a consumer.
My reasoning was that there might be edge cases where these assignments change during a benchmark run (e.g one consumer finishes early, starts late, etc).
That might not be needed in practice, though. WDYT?

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -333,11 +452,86 @@ public void stop(Platform platform) throws Exception {
+        ConsumerRecords<byte[], byte[]> poll() {
+            this.consumerLock.lock();
+            try {
+                return consumer.poll(Duration.ofMillis(50));
```
50 ms is way too short for a poll interval.  We don't want to use up so much CPU.  If there's nothing to consume it should wait for at least a minute.

stanislavkozlovski replied:
That's how it before but good point. If we remove the lock we should go with this

stanislavkozlovski committed:
Small changes

Better use of clientId() method
Rename threadCount to threadsPerWorker
Improve ConsumeBenchSpec javadoc

cmccabe commented on tests/kafkatest/services/trogdor/consume_bench_workload.py:
```diff
@@ -21,7 +21,7 @@
     def __init__(self, start_ms, duration_ms, consumer_node, bootstrap_servers,
                  target_messages_per_sec, max_messages, active_topics,
-                 consumer_conf, common_client_conf, admin_client_conf, consumer_group=None):
+                 consumer_conf, common_client_conf, admin_client_conf, consumer_group=None, consumer_count=1):
```
Should be "threadsPerWorker"

cmccabe commented on tests/kafkatest/tests/core/consume_bench_test.py:
```diff
@@ -120,13 +143,62 @@ def test_consume_group_bench(self):
                                                 consumer_conf={},
                                                 admin_client_conf={},
                                                 common_client_conf={},
+                                                consumer_count=2,
```
Same (should be threads_per_worker)

cmccabe commented:
Thanks @stanislavkozlovski .  Looks good.  Should be ready to commit once the variable name is fixed in the python code.

stanislavkozlovski committed:
Rename consumer_count to threads_per_worker

stanislavkozlovski commented:
@cmccabe the JDK 8 tests passed and the JDK11 failures look like unrelated flaky tests
